src/main/scala/com/box/castle/core/committer/manager/BatchSizeManager.scala

```diff
@@ -0,0 +1,87 @@
+package com.box.castle.core.committer.manager
+
+import com.box.castle.core.common.{BoundedQueue, ReadSample}
+import com.box.castle.core.config.CommitterConfig
+import org.joda.time.Duration
+import org.slf4s.Logging
+
+
+/**
+  * Implements logic to track reads and compute delay for following reads so as to ensure we always read close to
+  * bufferSize of data from Kafka.
+  *
+  */
+class BatchSizeManager(committerConfig: CommitterConfig,
+                       bufferSize: Int) extends Logging {
+
+  private val samples = new BoundedQueue[ReadSample](committerConfig.samplingSlots)
+  private var consecutiveFullBuffers = 0
+
+  /**
+    * Tracks bytes read from kafka and the corresponding timestamp based on specified sampling rate
+    * Also keeps track of consecutive full buffer reads seen.
+    *
+    * @param bytesRead
+    * @param timestamp
+    */
+  def track(bytesRead: Int, timestamp: Long) = {
+    if (bytesRead >= bufferSize) {
+      // Update number of times we have seen consecutive full buffer reads
+      consecutiveFullBuffers += 1
+    }
+    else {
+      // Reset counter if no more consecutive full buffer reads seen
+      consecutiveFullBuffers = 0
+    }
+
+    if (!samples.isFull || timestamp - samples.last.timestamp >= committerConfig.samplingInterval.getMillis) {
+      // If queue is full then we add a new sample only after sampling interval
+      // The bytes read is cumulative here so the new sample adds on to the last samples bytesRead
+      // This makes computation of rate more efficient
+      if (samples.isEmpty)
+        samples.enqueue(ReadSample(bytesRead, timestamp))
+      else
+        samples.enqueue(ReadSample(samples.last.bytesRead + bytesRead, timestamp))
```

Comment:
You can replace this if statement with:
`val sample = ReadSample(samples.lastOption.getOrElse(zeroByteSample).bytesRead + bytesRead, timestamp)`
`samples.enqueue(sample)`

in the constructor declare: `val zeroByteSample = ReadSample(0, 0)`